File: frontend/src/engine/ForecastEngine.js
/**
 * ForecastEngine.js
 * 
 * Handles generation and publication of standard GB electricity market forecasts.
 * Models Demand, Wind, and Solar over 48 Settlement Periods (SPs).
 * Supports modes:
 *  - 'manual': Instructor draws/modifies curves.
 *  - 'auto': Uses stochastic models to generate forecasts with noise.
 *  - 'mixed': Auto runs but manual overrides lock it.
 */

class ForecastVersion {
    constructor({ author, mode, demand, wind, solar, confidence, note }) {
        this.id = 'v' + Date.now();
        this.author = author;
        this.mode = mode;
        this.timestamp = new Date().toISOString();
        this.demand = demand || Array(48).fill(0);
        this.wind = wind || Array(48).fill(0);
        this.solar = solar || Array(48).fill(0);
        this.confidence = confidence || Array(48).fill(0); // 1-sigma uncertainty band
        this.note = note || '';
    }
}

class ForecastEngine {
    constructor(sp_per_day = 48) {
        this.sp_per_day = sp_per_day;
        this.published = null;      // The current shared version
        this.history = [];          // Archive of published versions
        this.mode = 'auto';         // 'manual' | 'auto' | 'mixed'
        this.manual_lock = false;   // If true, auto-update is paused
        this.skill_level = 0.9;     // 0.0 to 1.0; defines accuracy in auto model

        // Default base parameters for generation
        this.params = {
            base_demand: 35000, // MW
            peak_multiplier: 1.3,
            wind_capacity: 25000,
            solar_capacity: 15000,
            noise_level: 0.03 // 3%
        };
    }

    // Set simulation operating mode
    setMode(newMode) {
        if (['manual', 'auto', 'mixed'].includes(newMode)) {
            this.mode = newMode;
            // If switching explicitly to manual, lock the auto overrides
            this.manual_lock = (newMode === 'manual');
        }
    }

    // Publish a custom/drawn forecast (Manual mode)
    createManual(author, demand_ts, wind_ts, solar_ts, confidence_ts = null, note = "") {
        if (this.mode === 'auto') {
            console.warn("ForecastEngine is in 'auto' mode. Switch to 'manual' or 'mixed' to publish manual versions.");
            // We still allow it but usually UI warns.
        }

        // In Mixed mode, a manual publish engages the lock
        if (this.mode === 'mixed') {
            this.manual_lock = true;
        }

        const version = new ForecastVersion({
            author,
            mode: 'manual',
            demand: demand_ts,
            wind: wind_ts,
            solar: solar_ts,
            confidence: confidence_ts || this._defaultConfidence(demand_ts),
            note
        });

        this._publish(version);
        return version;
    }

    // Publish automatically based on internal models (Auto mode)
    autoGenerate(noise_multiplier = 1.0) {
        const demand = this._modelDemand(noise_multiplier);
        const wind = this._modelWind(noise_multiplier);
        const solar = this._modelSolar(noise_multiplier);
        const confidence = demand.map(d => d * this.params.noise_level * (1.1 - this.skill_level));

        const version = new ForecastVersion({
            author: 'NESO_AI',
            mode: 'auto',
            demand,
            wind,
            solar,
            confidence,
            note: 'Auto-generated DA forecast'
        });

        this._publish(version);
        return version;
    }

    // Inject a sudden change (Shock event)
    injectShock(shock_type, modifier_pct) {
        if (!this.published) {
            console.warn("Cannot inject shock without a published baseline.");
            return;
        }

        // Deep copy current version
        const newDemand = [...this.published.demand];
        const newWind = [...this.published.wind];
        const newSolar = [...this.published.solar];

        // Apply modifier
        for (let i = 0; i < this.sp_per_day; i++) {
            if (shock_type === 'wind_drop') newWind[i] *= (1 + modifier_pct);
            if (shock_type === 'demand_spike') newDemand[i] *= (1 + modifier_pct);
        }

        const version = new ForecastVersion({
            author: 'Instructor (System Override)',
            mode: 'manual',
            demand: newDemand,
            wind: newWind,
            solar: newSolar,
            confidence: [...this.published.confidence],
            note: `SYSTEM SHOCK: ${shock_type}`
        });

        this._publish(version);
        return version;
    }

    // Expose specific version
    getVersion(id) {
        return this.history.find(v => v.id === id);
    }

    // Internal publisher
    _publish(version) {
        this.published = version;
        this.history.push(version);
        // Sliding window: keep only last 50 versions to prevent memory bloat
        if (this.history.length > 50) {
            this.history.shift();
        }
        // Ideally this emits an event that socket/MarketEngine listens to.
        console.log(`[ForecastEngine] Published new version ${version.id} by ${version.author}`);
    }

    generateInitialDraft(currentSp) {
        return {
            demand: this._modelDemand(1.0),
            wind: this._modelWind(1.0),
            solar: this._modelSolar(1.0),
            margin: Array(48).fill(4000)
        };
    }

    // --- Internal Stochastic Models (per SP) ---

    _modelDemand(noise_mult) {
        // Basic double peak (morning, evening)
        const curve = [];
        for (let sp = 0; sp < this.sp_per_day; sp++) {
            let val = this.params.base_demand;
            const hour = sp / 2;

            // Morning peak (07:00 - 09:00)
            if (hour >= 7 && hour <= 9) val *= 1.15;
            // Evening peak (17:00 - 19:00)
            if (hour >= 17 && hour <= 19) val *= this.params.peak_multiplier;
            // Night drop (00:00 - 05:00)
            if (hour >= 0 && hour <= 5) val *= 0.7;

            const noise = 1 + (Math.random() - 0.5) * this.params.noise_level * noise_mult;
            curve.push(+(val * noise).toFixed(1));
        }
        return curve;
    }

    _modelWind(noise_mult) {
        const curve = [];
        // Wind often correlated, AR(1) proxy
        let current_yield = 0.4 + (Math.random() * 0.2); // 40-60% base
        for (let sp = 0; sp < this.sp_per_day; sp++) {
            current_yield += (Math.random() - 0.5) * 0.1 * noise_mult; // walk
            current_yield = Math.max(0, Math.min(1, current_yield)); // clamp
            curve.push(+(this.params.wind_capacity * current_yield).toFixed(1));
        }
        return curve;
    }

    _modelSolar(noise_mult) {
        const curve = [];
        for (let sp = 0; sp < this.sp_per_day; sp++) {
            const hour = sp / 2;
            let val = 0;
            if (hour > 6 && hour < 18) {
                // Bell curve peaking at 12:00
                const dist = Math.abs(12 - hour);
                val = Math.max(0, 1 - (dist / 6));
            }
            const noise = 1 + (Math.random() - 0.5) * (this.params.noise_level * 2) * noise_mult;
            curve.push(+(this.params.solar_capacity * val * noise).toFixed(1));
        }
        return curve;
    }

    _defaultConfidence(demand) {
        return demand.map(d => +(d * this.params.noise_level).toFixed(1));
    }
}

export default ForecastEngine;
export { ForecastVersion };

File: backend/engine/forecast_engine.py
"""
GridForge Forecast Engine — Python port of src/engine/ForecastEngine.js

Handles generation and publication of standard GB electricity market forecasts.
Models Demand, Wind, and Solar over 48 Settlement Periods (SPs).
Supports modes: manual, auto, mixed.
"""

from __future__ import annotations
import random
import math
from datetime import datetime


class ForecastVersion:
    def __init__(self, author: str, mode: str, demand: list[float], wind: list[float],
                 solar: list[float], confidence: list[float] | None = None, note: str = ""):
        self.id = f"v{int(datetime.now().timestamp() * 1000)}"
        self.author = author
        self.mode = mode
        self.timestamp = datetime.now().isoformat()
        self.demand = demand or [0.0] * 48
        self.wind = wind or [0.0] * 48
        self.solar = solar or [0.0] * 48
        self.confidence = confidence or [0.0] * 48
        self.note = note

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "author": self.author,
            "mode": self.mode,
            "timestamp": self.timestamp,
            "demand": self.demand,
            "wind": self.wind,
            "solar": self.solar,
            "confidence": self.confidence,
            "note": self.note,
        }


class ForecastEngine:
    def __init__(self, sp_per_day: int = 48, seed: int | None = None):
        self.sp_per_day = sp_per_day
        self.seed = seed  # stored so callers can persist + restore for determinism
        self._rng = random.Random(seed)  # isolated PRNG — never touches global random state
        self.published: ForecastVersion | None = None
        self.history: list[ForecastVersion] = []
        self.mode = "auto"
        self.manual_lock = False
        self.skill_level = 0.9

        self.params = {
            "base_demand": 35000,
            "peak_multiplier": 1.3,
            "wind_capacity": 25000,
            "solar_capacity": 15000,
            "noise_level": 0.03,
        }

    def set_mode(self, new_mode: str) -> None:
        if new_mode in ("manual", "auto", "mixed"):
            self.mode = new_mode
            self.manual_lock = new_mode == "manual"

    def create_manual(self, author: str, demand_ts: list[float], wind_ts: list[float],
                      solar_ts: list[float], confidence_ts: list[float] | None = None,
                      note: str = "") -> ForecastVersion:
        if self.mode == "mixed":
            self.manual_lock = True

        version = ForecastVersion(
            author=author,
            mode="manual",
            demand=demand_ts,
            wind=wind_ts,
            solar=solar_ts,
            confidence=confidence_ts or self._default_confidence(demand_ts),
            note=note,
        )
        self._publish(version)
        return version

    def auto_generate(self, noise_multiplier: float = 1.0) -> ForecastVersion:
        demand = self._model_demand(noise_multiplier)
        wind = self._model_wind(noise_multiplier)
        solar = self._model_solar(noise_multiplier)
        confidence = [round(d * self.params["noise_level"] * (1.1 - self.skill_level), 1) for d in demand]

        version = ForecastVersion(
            author="NESO_AI",
            mode="auto",
            demand=demand,
            wind=wind,
            solar=solar,
            confidence=confidence,
            note="Auto-generated DA forecast",
        )
        self._publish(version)
        return version

    def inject_shock(self, shock_type: str, modifier_pct: float) -> ForecastVersion | None:
        if not self.published:
            return None

        new_demand = list(self.published.demand)
        new_wind = list(self.published.wind)
        new_solar = list(self.published.solar)

        for i in range(self.sp_per_day):
            if shock_type == "wind_drop":
                new_wind[i] *= (1 + modifier_pct)
            if shock_type == "demand_spike":
                new_demand[i] *= (1 + modifier_pct)

        version = ForecastVersion(
            author="Instructor (System Override)",
            mode="manual",
            demand=new_demand,
            wind=new_wind,
            solar=new_solar,
            confidence=list(self.published.confidence),
            note=f"SYSTEM SHOCK: {shock_type}",
        )
        self._publish(version)
        return version

    def get_version(self, version_id: str) -> ForecastVersion | None:
        return next((v for v in self.history if v.id == version_id), None)

    def generate_initial_draft(self, current_sp: int = 1) -> dict:
        return {
            "demand": self._model_demand(1.0),
            "wind": self._model_wind(1.0),
            "solar": self._model_solar(1.0),
            "margin": [4000.0] * 48,
        }

    def _publish(self, version: ForecastVersion) -> None:
        self.published = version
        self.history.append(version)
        if len(self.history) > 50:
            self.history.pop(0)

    def _model_demand(self, noise_mult: float) -> list[float]:
        curve = []
        for sp in range(self.sp_per_day):
            val = self.params["base_demand"]
            hour = sp / 2
            if 7 <= hour <= 9:
                val *= 1.15
            if 17 <= hour <= 19:
                val *= self.params["peak_multiplier"]
            if 0 <= hour <= 5:
                val *= 0.7
            noise = 1 + (self._rng.random() - 0.5) * self.params["noise_level"] * noise_mult
            curve.append(round(val * noise, 1))
        return curve

    def _model_wind(self, noise_mult: float) -> list[float]:
        curve = []
        current_yield = 0.4 + self._rng.random() * 0.2
        for sp in range(self.sp_per_day):
            current_yield += (self._rng.random() - 0.5) * 0.1 * noise_mult
            current_yield = max(0, min(1, current_yield))
            curve.append(round(self.params["wind_capacity"] * current_yield, 1))
        return curve

    def _model_solar(self, noise_mult: float) -> list[float]:
        curve = []
        for sp in range(self.sp_per_day):
            hour = sp / 2
            val = 0.0
            if 6 < hour < 18:
                dist = abs(12 - hour)
                val = max(0, 1 - dist / 6)
            noise = 1 + (self._rng.random() - 0.5) * self.params["noise_level"] * 2 * noise_mult
            curve.append(round(self.params["solar_capacity"] * val * noise, 1))
        return curve

    def _default_confidence(self, demand: list[float]) -> list[float]:
        return [round(d * self.params["noise_level"], 1) for d in demand]
